baxter_collaboration_lib/src/robot_interface/gripper.cpp

```diff
@@ -146,17 +203,279 @@ bool Gripper::has_error()
 
 bool Gripper::is_sucking()
 {
+    if(type() != "suction") { capabilityWarning("is_sucking"); }
+
     // ROS_INFO("force is: %g\n",getGripperState().force);
     return getGripperState().position<80;
 }
 
 bool Gripper::is_gripping()
 {
+    return getGripperState().gripping==EndEffectorState::STATE_TRUE;
+}
+
+bool Gripper::hasForce()
+{
+    return getGripperProperties().controls_force == true;
+}
+
+bool Gripper::hasPosition()
+{
+    return getGripperProperties().controls_position == true;
+}
+
+bool Gripper::open(bool _block, double _timeout)
+{
+    if(type() == "electric")
+    {
+        return commandPosition(100.0, _block, _timeout);
+    }
+    else if (type() == "suction")
+    {
+        // check if the gripper is already not sucking
+        if(not is_sucking())
+        {
+            ROS_WARN("%s gripper of type %s is already open",
+                      getGripperLimb().c_str(), type().c_str());
+
+            return false;
+        }
+
+        return stop(_block, _timeout);
+    }
+    else
+    {
+        // give a warning if not capable and return
+        capabilityWarning("open");
+        return false;
+    }
+}
+
+bool Gripper::close(bool _block, double _timeout)
+{
+    if(type() == "electric")
+    {
+        return commandPosition(0.0, _block, _timeout);
+    }
+    else if (type() == "suction")
+    {
+        // no checks here for is_sucking() so that the suction time may be extended as necessary
+        return commandSuction(_block, _timeout);
+    }
+    else
+    {
+        // give a warning if not capable and return
+        capabilityWarning("close");
+        return false;
+    }
+}
+
+bool Gripper::commandPosition(double _position, bool _block, double _timeout)
+{
+    // give a warning if not capable and return
+    if(type() != "electric")
+    {
+        capabilityWarning("commandPosition");
+        return false;
+    }
+
+    // calibrate the electric gripper if needed
+    // do it in a blocking manner so that the gripper does not move before calibration
+    if(not is_calibrated())
+    {
+        ROS_INFO("Calibrating gripper. Please wait...");
+        calibrate(true, 2.0);
```

Comment:
oh i included the 2 second timeout here so that for e.g. the closing of the gripper doesn't conflict with the calibrate. but i can reduce the timeout if you like